### PR TITLE
[skip changelog] Document the programmers system of Arduino platforms

### PR DESCRIPTION
**Pull request by per1234**
**Please check if the PR fulfills these requirements**

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

---
* **What kind of change does this PR introduce?**

Docs update

---
- **What is the current behavior?**

The system of support for external programmers in Arduino platforms, although complex and in widespread use, is essentially undocumented.

---
* **What is the new behavior?**

The programmers features of Arduino platforms are fully documented:

- platform.txt
- "Burn Bootloader"/`erase` and `bootloader` actions
- "Upload using Programmer/`program` action
- Programmer inheritance from referenced platforms

---
- **Does this PR introduce a breaking change?**

No
* **Other information**:

This pull request takes into account the behavior of the Arduino IDE both before and after the available programmers were limited to the platform of the selected board (https://github.com/arduino/Arduino/pull/9900)

---
NOTE: the example programmer snippet added in https://github.com/arduino/arduino-cli/pull/925/commits/fc251822564ad704974cb38fad33aec07a8c54eb is missing some of the properties that are defined in [the real programmer definition in the Arduino AVR Boards platform](https://github.com/arduino/ArduinoCore-avr/blob/1.8.3/programmers.txt#L44-L51). After a lot of investigation, experimentation, and even consulting with the original author of the code, I was still unable to find any use of these properties in the current or any past IDE version that supports the 1.5 platform format. If these properties have no purpose, they only represent a source of confusion and maintenance burden for platform developers.

### CODE REVIEW

File: docs/platform-specification.md [hunk 1]
```diff
@@ -701,11 +739,53 @@ The file component of the port's path (e.g., `ttyACM0`) is available as the conf
...
+    usbasp.program.tool=avrdude
+    [......]
+    arduinoasisp.program.tool=avrdude
+    [......]
+
+This action can use the same [upload verification preference system](#upload-verification) as the `upload` action, via
+the **program.verify** property.
+
+When using the Arduino IDE, if the selected programmer is from a different platform than the board, the `program` recipe
+of the programmer's platform is used without overrides from the properties of the [platform.txt](#platformtxt) of the
```
Comment (ubidefeo) [inline docs/platform-specification.md:759]: these 3 "of" in the same paragraph sound a little off to me
"of the programmer's platform
"of the [platform.txt]"
"of the [board platform]"

**Review by ubidefeo — COMMENTED** [review]
beside the comment about the "of" the rest all looks good to me

File: docs/platform-specification.md (hunk 1 above)
Comment (per1234) [reply docs/platform-specification.md:759]: This is an excellent point. If I understand correctly, the problem is really about the nested "of"s in this part of the sentence:
>... the properties **of** the platform.txt **of** the board platform ...

Other than "from", which is not an option due to already being in use in the sentence, I didn't come up with a good replacement for the second "of". So my proposal is to replace the first of those two "of"s, as well as its parallel occurence earlier in the sentence, with "defined by":

>... the `program` recipe **defined by** the programmer's platform is used without overrides **from** the properties **defined by** the platform.txt **of** the board platform ...

What do you think @ubidefeo?

Comment (ubidefeo) [reply docs/platform-specification.md:759]: now I like it 👌🏼

Comment (per1234) [reply docs/platform-specification.md]: I have now made the suggested change:
https://github.com/arduino/arduino-cli/compare/a7502c9b636f88ab2aa81d8eb4fa8869653a3f3d..8a5b4f5c6a08f78d0ba69277923dcc19478b49e4
Thanks!

This is a complex paragraph about something that has regularly caused confusion for platform authors over the years. Although the consistent behavior provided by Arduino CLI is a significant improvement, it did make the documentation more difficult due to the added requirement to explain the differences in the behavior of the tools. So I think that the improvement to the readability is especially important here.

**Review by rsora — APPROVED** [review]

**Pull request merged**